src/index.ts

```diff
@@ -201,13 +201,14 @@ class SmartQueue {
       return;
     }
 
-    debug('Executing queue item', nextItem.item.id);
+    this.heat();
 
-    const startTime = Date.now();
+    debug('Executing queue item', nextItem.item.id);
 
     try {
-      const data = await nextItem.item.request(retryFn);
-
+      const requestPromise = nextItem.item.request(retryFn);
+      this.execute();
+      const data = await requestPromise;
```

Comment:
Won't it work the same if we just do 
```
this.execute();
const data = await nextItem.item.request(retryFn);
```
I guess as soon as we use async await this function will just be called in microtask and only after `await this.shift`. What do you think?

Reply:
It will work, but it will cause the cool down period to start a bit earlier than it should. as the actual request to the "rate limited api" only happens after you call the request function.
I think my way is a more safeguard approach, but it is a mili second to either side, so if you prefer code cleanness, we can go with your way

Reply:
okay, let's do it your way